src/library/Cardano/CoinSelection.hs

```diff
@@ -55,93 +164,62 @@ import GHC.Generics
 -- for all of the outputs, and a /remaining UTxO set/ from which all spent
 -- values have been removed.
 --
-newtype CoinSelectionAlgorithm i o u m e = CoinSelectionAlgorithm
+newtype CoinSelectionAlgorithm i o m e = CoinSelectionAlgorithm
     { selectCoins
         :: CoinSelectionOptions i o e
-        -> NonEmpty (Output o)
-        -> UTxO u
-        -> ExceptT (CoinSelectionError e) m (CoinSelection i o, UTxO u)
+        -> CoinMap i
+        -> CoinMap o
+        -> ExceptT (CoinSelectionError e) m (CoinSelection i o, CoinMap i)
     }
 
 -- | Represents the result of running a /coin selection algorithm/.
 --
 -- See 'CoinSelectionAlgorithm'.
 --
 data CoinSelection i o = CoinSelection
-    { inputs :: [Input i]
+    { inputs :: CoinMap i
       -- ^ A /subset/ of the original 'UTxO' that was passed to the coin
       -- selection algorithm, containing only the entries that were /selected/
       -- by the coin selection algorithm.
-    , outputs :: [Output o]
+    , outputs :: CoinMap o
       -- ^ The original set of output payments passed to the coin selection
       -- algorithm, whose total value is covered by the 'inputs'.
     , change :: [Coin]
       -- ^ A set of change values to be paid back to the originator of the
       -- payment.
-    } deriving (Generic, Show, Eq)
+    }
+    deriving (Generic, Show, Eq)
 
--- NOTE:
---
--- We don't check for duplicates when combining selections because we assume
--- they are constructed from independent elements.
---
--- As an alternative to the current implementation, we could 'nub' the list or
--- use a 'Set'.
```

Comment:
This assumption is no longer necessary.

However, we should test that:
- [ ] combining selections preserves all keys.
- [ ] combining selections preserves total coin value.

Reply:
Fixed by the following commits:

45f7021cea5ddaa76ead1c0f3777d3e5413fc11f
2f695cb86b966fafab573644b7f179b37d0c3184